### PR TITLE
Remove Database Isolation feature flag and run DB isolation tests

### LINKED ISSUE #41069: Make sure all tests are passing in DB Isolation mode
We have a Draft PR where we run all the DB tests in DB isolation mode.  The https://github.com/apache/airflow/pull/41067 is here and we should make sure we extract and fix all the problems there before we can say "enable AIP-44"

### PR DESCRIPTION
**Pull request by potiuk**
This PR removes AIP-44 feature flag and replaces "in-progress-disabled" test with dedicated "DatabaseIsolation" one.

The DatabaseIsolation test will run all "db-tests" with enabled DB isolation mode and running `internal-api` component - groups of tests marked with "skip-if-database-isolation" will be skipped.

 - [ ] tests/models/test_taskinstance.py (32 tests failing)
 - [ ] tests/operators/test_python.py (2 tests failing)
 - [ ] tests/sensors/test_external_task_sensor.py (2 tests failing)
 
Some of those tests might need to be excluded as they make no sense to be run (for example if those are tests that are testing scheduler or webserver code). It might also be that we need to improve our test harness for isolation mode in some cases. 

----

How to run the tests? It's fairly easy to run the tests with breeeze (you can checkout this pr with `gh pr checkout 41067` and work on top of it.

You can run tests using DB isolation already using database isolation. It works in the way that you need to enter breeze shell --database-isolation, split the terminals with tmux, run internal-api component in one of them and then you can run pytest tests in the other terminal. Again - sources are mounted to inside the container so you can easily iterate on tests (ctrl-c restart internal-api to pick up changes).

See: https://github.com/apache/airflow/blob/main/contributing-docs/testing/unit_tests.rst#running-tests-with-database-isolation for detailed description.

One tests that fully works for now is dag_processing/test_job_runner.py for example and a good way to see how it works - you will see that the test initializes DB direcly, but then the dag_processing code calls the internal-api server and you will see messages going back-forth in the "internal-api" output.

You should make sure that you use `pytest --run-db-tests-only` flag - and all non-db tests will be skipped, that should speed up your tests when you run whole module or directory.

----- 

How to submit fixes?

After you make a fix - submit a PR explaining which tests are fixed with `Related: #41067` in the commit message, That's it. I will be continuously rebasing this PR on top of main and keeping the inventory of what has been fixed. 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
This PR will have quite a number of failing tests still - some of them because they should just be disabled - on a module level in most cases  (with adding `pytest.mark.skuip-if-database-isolation` to `pytest.mark.db-test`. 

Some of them because there are still some missing pieces to fix in the internal API. 

The "special"  `DatabaseIsolation` test that replaces "InProgressDisabled" runs the special "Database Isolation" mode that allows all our test fixtures and generally "test code" to access the database, while any code that is being tested from "airflow" should get special TracingSession assigned that will fail if the "airflow" code will attempt to use the database directly rather than via internal API.

I will make an extra pass today - once it completes - and disable all the "big" modules that should be disabled from testing, and maybe fix some obvious cases. But anyone (especially @jscheffl @vincbeck and @dstandish ) are most welcome to sign-up to fix some of those in parallel (please comment here when you pick something - I will be merging my fixes individually as separate PRs and will keep on rebasing that PR until we will get it hopefully green before 2.10rc1.

**Comment by aritra24** [comment]
I'll try taking a look at some failures tomorrow my time

**Comment by jscheffl** [comment]
I LOVE pull-requests which delete more lines than adding. But as being DRAFT, looking forward for review (if family time permits tomorrow)

**Comment by potiuk** [comment]
Ok. I think this one is ready for anyone else to help - we have quite a number of failing tests, but a lot of them can be likely easily fixed. I added the summary of the failing tests in the issue description.

I also added short instructions on how to run the tests. So anyone willing to help, feel free to comment here and say which one you are looking at, and make PR out of your fixes. I will be continuously rebasing this one until it gets - hopefully - green.

**Comment by jscheffl** [comment]
> Ok. I think this one is ready for anyone else to help - we have quite a number of failing tests, but a lot of them can be likely easily fixed. I added the summary of the failing tests in the issue description.
> 
> I also added short instructions on how to run the tests. So anyone willing to help, feel free to comment here and say which one you are looking at, and make PR out of your fixes. I will be continuously rebasing this one until it gets - hopefully - green.

Note: Probably only few get to know about this P as it is Draft and you get notifications only for "Real" PRs?

**Comment by potiuk** [comment]
> Note: Probably only few get to know about this PR as it is Draft and you get notifications only for "Real" PRs?

Not for most of the reviewers - because I deliberately created it as "fulll" PR and only after that marked it as "Draft" 

But yeah - I added Elad and Nicolas after that - and they are not notified - see the image (the crescent moon is sign of not being notified). Let me mark it as read and back to draft. That should fix it.

image

**Comment by jscheffl** [comment]
I was wondering and debugging a while for the one failed test in `tests/core/test_sentry.py` and came to the point that the decorator `@time_machine.travel(CRUMB_DATE)` on this test modifies the virtual time... and thus the authentication token is created also relative to `CRUMB_DATE` == May 2019... and when testing is expired for a few years at internal API.

Any idea how to fix this?

**Comment by potiuk** [comment]
Probably we can add option ti disable authentication check for unit tests ?

**Comment by jscheffl** [comment]
> Probably we can add option to disable authentication check for unit tests ?

Mhm, I see the next vulnerability coming :-D

**Comment by jscheffl** [comment]
@potiuk I analyzed the failure in `tests/always/test_secrets_backends.py`  -reason for the failure is that the Internal API server misses the `AIRFLOW__CORE__FERNET_KEY` - I inspected the code and do not understand "how" I could get this ENV in Breeze into the API Server (not creating a new one, must be matching the Fernet Key used by Pytest as well) - any hint welcome. Then this error would be fixed (maybe others as side effect as well)

**Comment by potiuk** [comment]
> @potiuk I analyzed the failure in `tests/always/test_secrets_backends.py` -reason for the failure is that the Internal API server misses the `AIRFLOW__CORE__FERNET_KEY` - I inspected the code and do not understand "how" I could get this ENV in Breeze into the API Server (not creating a new one, must be matching the Fernet Key used by Pytest as well) - any hint welcome. Then this error would be fixed (maybe others as side effect as well)

https://github.com/apache/airflow/pull/41088 fixes it

**Comment by aritra24** [comment]
I'm looking at some of the test_python failures, should have something by EoD

**Comment by potiuk** [comment]
I am **very little** available today - but if somoene wants to take a stab at some of those - feel free, I will try to look them up. For now **some** tests are hanging - I will have to take a look, but tomorrow I will try to make it completes and see how many tests are left - as some of the fixes fixed more than one module I **think** (the FERNET_KEY fixed by #41088 I saw in a few places)

**Comment by vincbeck** [comment]
I'll try to pick some tests when I can, possibly around the end of the week

**Comment by vincbeck** [comment]
I tried to run `tests/datasets/test_dataset.py` and it seems all tests are passing?

**Comment by aritra24** [comment]
I was trying to figure out test_python. It looks like the dags are not getting committed?
```
[2024-07-30T09:53:52.608+0000] {taskinstance.py:2735} INFO - Id is dag_testpythonoperator and in dagbag is []
[2024-07-30T09:53:52.611+0000] {rpc_api_endpoint.py:160} ERROR - Error executing method 'airflow.models.taskinstance.TaskInstance._check_and_change_state_before_execution'. error_id=bfb26504-0e80-4a21-b3c9-1e4e8a392635
Traceback (most recent call last):
  File "/opt/airflow/airflow/api_internal/endpoints/rpc_api_endpoint.py", line 231, in internal_airflow_api
    output = handler(**params, session=session)
  File "/opt/airflow/airflow/api_internal/internal_api_call.py", line 135, in wrapper
    return func(*args, **kwargs)
  File "/opt/airflow/airflow/utils/session.py", line 94, in wrapper
    return func(*args, **kwargs)
  File "/opt/airflow/airflow/models/taskinstance.py", line 2737, in _check_and_change_state_before_execution
    task_instance.task = dag.task_dict[ti.task_id]
AttributeError: 'NoneType' object has no attribute 'task_dict'
```
 I saw this pr #41082 added some session.commits but I don't think that was enough. I am trying to look a bit more on what I could do. I tried tinkering around in conftest.py's init and create_dagrun methods but that didn't do much. Will see if I can find anything more. @potiuk if you have any directions you think might be good to look at do let me know.

**Comment by potiuk** [comment]
OK. The tests are not hanging any more - and I added semi-automated way of retrieving the list of remaining test and will keep on refreshing it.

**Comment by vincbeck** [comment]
> OK. The tests are not hanging any more - and I added semi-automated way of retrieving the list of remaining test and will keep on refreshing it.

Weird, `tests/datasets/test_dataset.py` is succeeding on my side. Maybe it is failing when running with others and, somehow, they are conflicting each other? But when I run `pytest tests/datasets/test_dataset.py`, it is succeeding

**Comment by potiuk** [comment]
> Weird, tests/datasets/test_dataset.py is succeeding on my side. Maybe it is failing when running with others and, somehow, they are conflicting each other? But when I run pytest tests/datasets/test_dataset.py, it is succeeding

Rebase. Those are new dataset alias tests that are failing

**Comment by potiuk** [comment]
We are racing with time :(

**Comment by jscheffl** [comment]
> We are racing with time :(

Agree. There are a hell-lot of fixes open. Good thing that I saw is that the function is general working, all I saw are environment or test issues in DB sync between pytest/backend.

Would it make sense to skip the failing tests and get this to main and follow-up replacing skips incrementally later?

**Comment by potiuk** [comment]
> Would it make sense to skip the failing tests and get this to main and follow-up replacing skips incrementally later?

I'd say there are still some **real** issues there - and I think there are still few improvements to make in the harness, I think we can spoil much more for others than they for us to be honest - because they will have to deal also with the test environment failing in new tests

**Comment by jscheffl** [comment]
FYI I attempted to fix tests/datasets/test_manager.py but failed as the "task_instance mock is "wrongly" serialized and therefore the internal API failing in the backend.
To fix this a "real" TaskInstance object would be needed that is correctly serialized... What is the "fastest" options making a mock?

**Comment by potiuk** [comment]
Updated the current state

**Comment by vincbeck** [comment]
I think there is a bug in the `TracebackSessionForTests` class. See error below:

```
self = <airflow.settings.TracebackSessionForTests object at 0xffff8f812c70>, item = 'scalar'

    def __getattr__(self, item):
        test_code, frame_summary = self.is_called_from_test_code()
        if test_code:
            return getattr(self.current_db_session, item)
        print("POUETTT")
        print(frame_summary)
        raise RuntimeError(
>           "TracebackSessionForTests object was used but internal API is enabled. "
            "Only test code is allowed to use this object.\n"
            f"Called from:\n    {frame_summary.filename}: {frame_summary.lineno}{frame_summary.colno}\n"
            f"     {frame_summary.line}\n\n"
            "You'll need to ensure you are making only RPC calls with this object. "
            "The stack list below will show where the TracebackSession object was called:\n"
            + get_cleaned_traceback(self.traceback)
            + "\n\nThe stack list below will show where the TracebackSession object was created:\n"
            + get_cleaned_traceback(self.created_traceback)
        )
E       AttributeError: 'FrameSummary' object has no attribute 'colno'
```

**Comment by potiuk** [comment]
> I think there is a bug in the TracebackSessionForTests class.

Yep. saw it too. Apparently some frames have no colno. Removed it and pushed.

**Comment by vincbeck** [comment]
I think there is another one, I am looking into it but you'll probably be faster than me. I think there is something off in `is_called_from_test_code`, some sessions are created in tests but `is_called_from_test_code` returns `False`. Example:

```
ERROR tests/providers/amazon/aws/hooks/test_chime.py::TestChimeWebhookHook::test_build_chime_payload_message_length - RuntimeError: TracebackSessionForTests object was used but internal API is enabled. Only test code is allowed to use this object.
Called from:
    /opt/airflow/airflow/utils/db.py: 132
     if not session.scalar(select(1).where(conn.__class__.conn_id == conn.conn_id)):

You'll need to ensure you are making only RPC calls with this object. The stack list below will show where the TracebackSession object was called:
  File "/usr/local/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/opt/airflow/tests/providers/amazon/aws/hooks/test_chime.py", line 46, in setup_method
    db.merge_conn(
  File "/opt/airflow/airflow/utils/session.py", line 97, in wrapper
    return func(*args, session=session, **kwargs)
  File "/opt/airflow/airflow/utils/db.py", line 132, in merge_conn
    if not session.scalar(select(1).where(conn.__class__.conn_id == conn.conn_id)):


The stack list below will show where the TracebackSession object was created:
  File "/usr/local/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/opt/airflow/tests/providers/amazon/aws/hooks/test_chime.py", line 46, in setup_method
    db.merge_conn(
  File "/opt/airflow/airflow/utils/session.py", line 96, in wrapper
    with create_session() as session:
  File "/usr/local/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
```

**Comment by vincbeck** [comment]
Probably the issue is here?

```
        for tb in airflow_frames[::-1]:
            if tb.filename.startswith(AIRFLOW_PATH):
                if tb.filename.startswith(AIRFLOW_TESTS_PATH):
                    # this is a session created directly in the test code
                    return True, None
                else:
                    # We should not do that?
                    ------>return False, tb
```

**Comment by potiuk** [comment]
> I think there is another one, I am looking into it but you'll probably be faster than me. I think there is something off in is_called_from_test_code, some sessions are created in tests but is_called_from_test_code returns False:

Yes. The detection of the test code vs "tested" code is somewhat imperfect and arbitrary - we simply guess that given code is called as test setup/teardown. In many cases it is good (fixtures, calling operator "run" command. But some tests might indeed run directly a db calls to setup something before or cleanup after.

I wonder if we should maybe change the strategy  here and possibly rather than try to run "all tests" in isolation mode, maybe we should just run a select subset of them. But it would be a bit difficult to choose the right subset. 

WDYT? 

For now excluding those tests from db_isolation is easiest

**Comment by vincbeck** [comment]
> > I think there is another one, I am looking into it but you'll probably be faster than me. I think there is something off in is_called_from_test_code, some sessions are created in tests but is_called_from_test_code returns False:
> 
> Yes. The detection of the test code vs "tested" code is somewhat imperfect and arbitrary - we simply guess that given code is called as test setup/teardown. In many cases it is good (fixtures, calling operator "run" command. But some tests might indeed run directly a db calls to setup something before or cleanup after.
> 
> I wonder if we should maybe change the strategy here and possibly rather than try to run "all tests" in isolation mode, maybe we should just run a select subset of them. But it would be a bit difficult to choose the right subset.
> 
> WDYT?
> 
> For now excluding those tests from db_isolation is easiest

I was thinking, does it make sense to run providers tests in DB isolation mode? We might want to skip them all?

**Comment by vincbeck** [comment]
> > I think there is another one, I am looking into it but you'll probably be faster than me. I think there is something off in is_called_from_test_code, some sessions are created in tests but is_called_from_test_code returns False:
> 
> Yes. The detection of the test code vs "tested" code is somewhat imperfect and arbitrary - we simply guess that given code is called as test setup/teardown. In many cases it is good (fixtures, calling operator "run" command. But some tests might indeed run directly a db calls to setup something before or cleanup after.
> 
> I wonder if we should maybe change the strategy here and possibly rather than try to run "all tests" in isolation mode, maybe we should just run a select subset of them. But it would be a bit difficult to choose the right subset.
> 
> WDYT?
> 
> For now excluding those tests from db_isolation is easiest

I created a draft PR to suggest a fix: #41193

**Comment by potiuk** [comment]
> I was thinking, does it make sense to run providers tests in DB isolation mode? We might want to skip them all? 

I actually thought it's a good idea - since we can find if our providers are doing someething like calling the DB  in their code directly. Note- that we are only running db_tests (which is already quite a subset of all tests of providers - we are not running all "pure" unit tests that do not require DB - so we are limiting provider tests already.

**Comment by vincbeck** [comment]
> > I was thinking, does it make sense to run providers tests in DB isolation mode? We might want to skip them all?
> 
> I actually thought it's a good idea - since we can find if our providers are doing someething like calling the DB in their code directly. Note- that we are only running db_tests (which is already quite a subset of all tests of providers - we are not running all "pure" unit tests that do not require DB - so we are limiting provider tests already.

I see. I dont have much solution here then ^^

**Comment by jscheffl** [comment]
> > > I was thinking, does it make sense to run providers tests in DB isolation mode? We might want to skip them all?
> > 
> > 
> > I actually thought it's a good idea - since we can find if our providers are doing someething like calling the DB in their code directly. Note- that we are only running db_tests (which is already quite a subset of all tests of providers - we are not running all "pure" unit tests that do not require DB - so we are limiting provider tests already.
> 
> I see. I dont have much solution here then ^^

+1 for adding to the tests - if it blocks merging the PR I'd be okay to add it later but... yes especiually for custom operators in providers, they will break if not using internal API and iterating over the DB... or they might need to be marked as being non-compliant to databse isolation... but then would get a hard time to be migrated to Ariflow 3.0... so will be a good readiness check towards 3.0 compatability.

**Comment by potiuk** [comment]
I pushed a bit better detection mechanism for tested vs. testing code.  Should cover few more cases I saw as "testing" code

* running dag.create_dagrun() in tests directly
* running any `utils.db` in tests directly

```
        if any(filename.endswith("conftest.py") for filename, _, _, _ in airflow_frames):
            # This is a fixture call
            return True, None
        if len(airflow_frames) >= 2 and airflow_frames[-2].filename.startswith(AIRFLOW_TESTS_PATH):
            # Let's look at what we are calling directly from the test code
            current_filename, current_method_name = airflow_frames[-1].filename, airflow_frames[-1].name
            if (current_filename, current_method_name) in (
                (AIRFLOW_MODELS_BASEOPERATOR_PATH, "run"),
                (AIRFLOW_MODELS_DAG_PATH, "create_dagrun"),
            ):
                # This is baseoperator run method that is called directly from the test code and this is
                # usual pattern where we create a session in the test code to create dag_runs for tests.
                # If `run` code will be run inside a real "airflow" code the stack trace would be longer
                # and it would not be directly called from the test code. Also if subsequently any of the
                # run_task() method called later from the task code will attempt to execute any DB
                # method, the stack trace will be longer and we will catch it as "illegal" call.
                return True, None
            if current_filename == AIRFLOW_DB_UTILS_PATH:
                # This is a util method called directly from the test code
                return True, None
        for tb in airflow_frames[::-1]:
            if tb.filename.startswith(AIRFLOW_PATH):
                if tb.filename.startswith(AIRFLOW_TESTS_PATH):
                    # this is a session created directly in the test code
                    return True, None
                else:
                    return False, tb


```

**Comment by potiuk** [comment]
Updated latest list

**Comment by potiuk** [comment]
We have far less number of issues now to look at !

**Comment by potiuk** [comment]
BTW. I think the list is not **complete** yet -> because we see maximum 50 errrors per failed test group :(

**Comment by potiuk** [comment]
changed to 300 - let's hope there are not too many new ones :(

File: dev/breeze/src/airflow_breeze/utils/run_tests.py
Comment (jscheffl) [inline dev/breeze/src/airflow_breeze/utils/run_tests.py:333]: I think at least should be reverted before merge.

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/run_tests.py:333]: Or fixed to skip the core parts (that's what I plan to do) - but yes - it's temporary

**Comment by potiuk** [comment]
> Would you like an extra pair of hands on this one? I tested a couple of them and they are passing. I think they are already fixed. I am wondering if there are any leftovers


Sure. I  keep on rebasing and merging changes and the list above gets updated after every rebase. 

There are likely some side-effects unfortunately - even if some tests are passing individually, they might be impacted by other failing tests unfortunately - that's why some of these might show up as failing even if they pass individually :( 

But yeah - a pair of hands and eyes on some of those is most welcome

**Comment by bugraoz93** [comment]
I was focused on the changes and I totally missed this one opened #41290 for tests/operators/test_weekday.py. I should have mentioned it. :(

**Comment by jscheffl** [comment]
> I was focused on the changes and I totally missed this one opened #41290 for tests/operators/test_weekday.py. I should have mentioned it. :(

I am currently on the broken operators, TriggerDagRunOperator is my next PR... before you do parallel work... stay tuned..

**Comment by bugraoz93** [comment]
> I am currently on the broken operators, TriggerDagRunOperator is my next PR... before you do parallel work... stay tuned..

You were faster :) Yes, will stay tuned 👍 
I will check these next
* `tests/providers/cncf/kubernetes/operators/test_job.py`
* `tests/providers/cncf/kubernetes/operators/test_spark_kubernetes.py`

**Comment by jscheffl** [comment]
FYI, I am moving further attempting to fix the leftover "Other" failures.

**Comment by bugraoz93** [comment]
I am going to attempt to fix `tests/models` starting from `tests/models/test_variable.py`(will go from bottom to top).

**Comment by bugraoz93** [comment]
I am moving forward to `tests/models/test_trigger.py `, FYI

**Comment by jscheffl** [comment]
Okay, then I jump now in `tests/models/test_taskinstance.py` - which might generate some grair hair for me as it seems to be... huge!

**Comment by bugraoz93** [comment]
Yes, that’s definitely a huge one! Hope everything goes smoothly. 🤞 Let me know if I can help.

**Comment by bugraoz93** [comment]
I see some tests are still failing for the files I worked on. I will check them again today and will include the fixes into the last PR I am working on. They seemed they were passing. I will double-check the problems.

**Comment by vincbeck** [comment]
> I see some tests are still failing for the files I worked on. I will check them again today and will include the fixes into the last PR I am working on. They seemed they were passing. I will double-check the problems.

Sometimes some tests are passing individually but not when run in parallel with others (because of shared resources, others conflicts ...). It might be the reason (or might be not)

**Comment by bugraoz93** [comment]
> Sometimes some tests are passing individually but not when run in parallel with others (because of shared resources, others conflicts ...). It might be the reason (or might be not)

Yes, exactly, that's why I used the full parameters that we are using in the CI/CD steps. All the steps are using `--with-db-init` parameters. I thought this would separate the `pytest` commands due to initialising the db again for each command in the same CI/CD step. I am also suspecting similarly, even though the db initialises again, I think we have some remaining resources that are affecting the tests.
For example, while testing the providers, I have used this one which is just copy and paste from the job

`tests/providers --ignore=tests/providers/amazon --ignore=tests/providers/google --verbosity=0 --strict-markers --durations=100 --maxfail=1000 --color=yes --junitxml=/files/test_result-providers_-amazon_google-postgres.xml --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 --disable-warnings -rfEX --run-db-tests-only --ignore=tests/system --ignore=tests/integration --warning-output-path=/files/warnings-providers_-amazon_google-postgres.txt --ignore=helm_tests --with-db-init --no-cov`

**Comment by potiuk** [comment]
Rebased with latest changes to see where we are in main

**Comment by jscheffl** [comment]
I fix the follow-up of test_variable.py

**Comment by bugraoz93** [comment]
`tests/decorators/test_python.py` seems like an easy one. Creating the PR shortly

**Comment by potiuk** [comment]
OK @vincbeck @jscheffl @aritra24 @eladkal others - this one seem to be finally green !  Made it "ready for review" and once we merge it - we can attempt to get it into v2-10-test branch (we need to add a little documentation to it- but I'd say we should aim to either get it for rc2 of 2.10 if we have one or maybe aim to get it in for 2.10.1

**Review by vincbeck — APPROVED** [review]
Really nice work from this PR and the others already merged which fixed tests one by one! 🎉

**Review by aritra24 — APPROVED** [review]
Yay! This looks great! Amazing work all!

**Review by jscheffl — APPROVED** [review]
Wow. What a stretch. Cool Teamwork (felt like a bit of distributed hackaton) getting this green! I feel a bit exhaiusted after this fix series and LGTM... then I can build on top of this and make the WIP of AIP-69 to main (hopefully soon!)

**Comment by bugraoz93** [comment]
Awesome to see this green! Amazing work!

**Comment by ferruzzi** [comment]
Sorry I missed out on this one, looks like a cool project.    It had green tests and two approvals, but also over 200 "hidden items" in the discussion thread so I'm afraid to hit the merge button on it without more context.   Congrats on getting it done.

**Comment by jscheffl** [comment]
@potiuk good to merge or still hold this back?

**Comment by potiuk** [comment]
> @potiuk good to merge or still hold this back?

I am struggling with it a bit - because it will likely make it more difficult for contributors to add their tests to main - while we are anyhow targetting the AIP-44 to be ripped off from Airflow 3 eventually. 

Looking at the kind of issues we had to go through, it will be a bit of a pain to merge that one to `main` - only to remove it afterwards, so I'd rather cherry-pick and merge that one to v2-10-test, however technically speaking this is a "new" feature enabled there, so if we folow it literally, it should go to 2.11.0. Probably the same kind of issue that @uranusjr mentioned in the `[RESULT][VOTE]` discussion: https://lists.apache.org/thread/56b7bdrzszkqtd6kgy360okw72568ld2

So I **think** maybe we should have a v2-11-test branch already created and we should start cherry-picking thing there that do not belong to 2.10.0 (and periodically rebasing it - or use cherry-picker to simultaneously cherry-pick things to both v2-10-test and v2-11-test when cherrypicking stuff to v2-10-test.

@utkarsharma2 @ephraimbuddy @kaxil @uranusjr @jscheffl   - would love to hear your thoughts about it.

**Comment by ephraimbuddy** [comment]
> > @potiuk good to merge or still hold this back?
> 
> I am struggling with it a bit - because it will likely make it more difficult for contributors to add their tests to main - while we are anyhow targetting the AIP-44 to be ripped off from Airflow 3 eventually.
> 
> Looking at the kind of issues we had to go through, it will be a bit of a pain to merge that one to `main` - only to remove it afterwards, so I'd rather cherry-pick and merge that one to v2-10-test, however technically speaking this is a "new" feature enabled there, so if we folow it literally, it should go to 2.11.0. Probably the same kind of issue that @uranusjr mentioned in the `[RESULT][VOTE]` discussion: https://lists.apache.org/thread/56b7bdrzszkqtd6kgy360okw72568ld2
> 
> So I **think** maybe we should have a v2-11-test branch already created and we should start cherry-picking thing there that do not belong to 2.10.0 (and periodically rebasing it - or use cherry-picker to simultaneously cherry-pick things to both v2-10-test and v2-11-test when cherrypicking stuff to v2-10-test.
> 
> @utkarsharma2 @ephraimbuddy @kaxil @uranusjr @jscheffl - would love to hear your thoughts about it.

Yeah. This is a concern. Should we have a branch called v2-dev where all airflow 2 PRs will go into, then the release manager would cherry-pick to v2-10-test when releasing? I think that having v2-11-test might make us miss PRs or having multiple places to PR to

**Comment by potiuk** [comment]
> Yeah. This is a concern. Should we have a branch called v2-dev where all airflow 2 PRs will go into, then the release manager would cherry-pick to v2-10-test when releasing? I think that having v2-11-test might make us miss PRs or having multiple places to PR to

Shall we spin it off to a devlist discussion?

**Comment by jscheffl** [comment]
> > Yeah. This is a concern. Should we have a branch called v2-dev where all airflow 2 PRs will go into, then the release manager would cherry-pick to v2-10-test when releasing? I think that having v2-11-test might make us miss PRs or having multiple places to PR to
> 
> Shall we spin it off to a devlist discussion?

Devlist discussion might be fair.

Options that I see contributing above:

1. adding to main to then rip it out just to bring it to 2.x branch... beneficial would be at least having less differences between v2 and v3 for a moment... but benefit is small
2. adding to 2.10-test would be is you account it precise not a bugfix. But not merging it woul dmean that in 2.10 line the chance is high that the existing internal API is degrading very fast - that is why we had the tests added. Whereas the feature flag removal is just a feature flag removal. The PR is not adding functionality per-se but just enabling it.

My main concern and pain is the high probability that the internal API degrades very fast in 2.x line. 

How about... we split the PR into:
* Feature flag --> Airflow 2.11 to be cherry-picked - this will probably not have risk to degrade
* Internal API Tests --> Merge to v2-10-test to ensure API works - is just adding tests, not "features"

**Comment by potiuk** [comment]
> How about... we split the PR into:

> * Feature flag --> Airflow 2.11 to be cherry-picked - this will probably not have risk to degrade
> * Internal API Tests --> Merge to v2-10-test to ensure API works - is just adding tests, not "features"


I like this. Let's do it.

**Comment by potiuk** [comment]
cc: @utkarsharma2 - I am merging this one as with some of the test changes (only tests) I believe it brings stability to the running tests and you need it for Python client relase.

**Comment by potiuk** [comment]
> 
cc: @utkarsharma2 - I am merging this one as with some of the test changes (only tests) I believe it brings stability to the running tests and you need it for Python client relase.

Not this one - the other one for v2-10-test

**Comment by potiuk** [comment]
@jscheffl -> shall we close that one then after #41567  merged? Are you good with that we are not planning to merge it and essentially we will remove AIP-44 from main - is that fine for you ?

**Comment by utkarsharma2** [comment]
> cc: @utkarsharma2 - I am merging this one as with some of the test changes (only tests) I believe it brings stability to the running tests and you need it for Python client relase.

Thanks @potiuk, much appreciated :)

**Comment by jscheffl** [comment]
> @jscheffl -> shall we close that one then after #41567 merged? Are you good with that we are not planning to merge it and essentially we will remove AIP-44 from main - is that fine for you ?

I think to be realistic and as we are riding an almost dead horse here... yes. I can accept if we are not removing the feature flag in Airflow 2.11. Is too much discussion for a short living improvement. Attempting to make AIP-69 with Airflow 2.10 anyway will assume the feature flag must be set.

**Comment by potiuk** [comment]
We still **might** remove the feature flag in 2.11. This will be easy - we do not need to keep that PR around for that.

**Pull request closed without merging**
